xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -134,11 +154,794 @@ func (s) TestRingHash_ReconnectToMoveOutOfTransientFailure(t *testing.T) {
+		Routes:    []*v3routepb.RouteConfiguration{route},
+		Listeners: []*v3listenerpb.Listener{listener},
+	})
+	if err != nil {
+		t.Fatalf("failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Note each type of RPC contains a header value that would always be hashed
+	// to a specific backend as the header value matches the value used to
+	// create the entry in the ring. However, the regex rewrites all numbers to
+	// "foo", and header values only differ by numbers, so they all end up
+	// hashing to the same value.
+	gotPerBackend := make(map[string]int)
+	for _, backend := range backends {
+		ctx := metadata.NewOutgoingContext(ctx, metadata.Pairs("address_hash", backend.Address+"_0"))
+		res := checkRPCSendOK(t, ctx, client, 100)
+		for addr, count := range res {
+			gotPerBackend[addr] += count
+		}
+	}
+	if want := 1; len(gotPerBackend) != want {
+		t.Errorf("Got RPCs routed to %v backends, want %v", len(gotPerBackend), want)
+	}
+	for _, got := range gotPerBackend {
+		if want := 400; got != want {
+			t.Errorf("Got %v RPCs routed to a backend, want %v", got, want)
+		}
+	}
+
+}
+
+// computeIdealNumberOfRPCs computes the ideal number of RPCs to send so that
+// we can observe an event happening with probability p, and the result will
+// have value p with the given error tolerance.
+func computeIdealNumberOfRPCs(p, errorTolerance float64) int {
+	if p < 0 || p > 1 {
+		panic("p must be in (0, 1)")
+	}
+	numRPCs := math.Ceil(p*(1-p)) * 5. * 5. / errorTolerance / errorTolerance
+	return int(numRPCs + 1000.) // add 1k as a buffer to avoid flakyness.
+}
+
+// TestRingHash_NoHashPolicy tests that ring hash policy that hashes using a
+// random value.
+func (s) TestRingHash_NoHashPolicy(t *testing.T) {
+	backends, stop := startTestServiceBackends(t, 2)
+	defer stop()
+	numRPCs := computeIdealNumberOfRPCs(.5, errorTolerance)
+
+	xdsServer, nodeID, _, xdsResolver, stop := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{
+		AllowResourceSubset: true,
+	})
+	defer stop()
+	clusterName := "cluster"
+	endpoints := e2e.EndpointResourceWithOptions(e2e.EndpointOptions{
+		ClusterName: clusterName,
+		Localities: []e2e.LocalityOptions{{
+			Backends: backendOptions(t, backends),
+			Weight:   1,
+		}},
+	})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+	})
+	// Increasing min ring size for random distribution.
+	config := testutils.MarshalAny(t, &v3ringhashpb.RingHash{
+		HashFunction:    v3ringhashpb.RingHash_XX_HASH,
+		MinimumRingSize: &wrapperspb.UInt64Value{Value: 100000},
```

Comment:
Actually this doesn't work because raising the minimum ring size is gated behind an environment variable. I'll look into:
- whether increasing the ring size really makes the distribution more uniform
- how to set the environment variable in tests (e.g. via mutating envconfig perhaps?)

Reply:
I ended up setting envconfig in a helper here.